File: uhrwerk-backend-mysql/src/test/java/io/qimia/uhrwerk/dao/PartitionDependencyDAOTest.java
package io.qimia.uhrwerk.dao;

import io.qimia.uhrwerk.ConnectionHelper;
import io.qimia.uhrwerk.common.metastore.dependency.DependencyResult;
import io.qimia.uhrwerk.common.model.Partition;
import io.qimia.uhrwerk.common.model.PartitionUnit;
import org.junit.jupiter.api.Test;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.util.HashSet;

import static org.junit.jupiter.api.Assertions.*;

public class PartitionDependencyDAOTest {

    Connection db;

    @org.junit.jupiter.api.BeforeEach
    void setUp() throws SQLException {
        db = ConnectionHelper.getConnection();
    }

    @org.junit.jupiter.api.AfterEach
    void tearDown() throws SQLException {
        var deletePartDependencies = db.createStatement();  // In case of some lost source data
        deletePartDependencies.execute("DELETE FROM PARTITION_DEPENDENCY");
        deletePartDependencies.close();
        var deletePartitions = db.createStatement();
        deletePartitions.execute("DELETE FROM PARTITION_");
        deletePartitions.close();
        if (db != null) if (!db.isClosed()) db.close();
    }

    Partition[] setupPartitions(LocalDateTime ts) throws SQLException {
        String insertQuery = "INSERT INTO PARTITION_(id, target_id, partition_ts)\n"
                + "VALUES(?,?,?)";

        // WARNING: uses fake target ids
        long[] fakeIds = {123L, 234L, 345L};
        Partition[] out = new Partition[fakeIds.length];
        for (int i = 0; i < fakeIds.length; i++) {
            var p = new Partition();
            p.setTargetId(fakeIds[i]);
            p.setPartitionTs(ts);
            p.setPartitionUnit(PartitionUnit.MINUTES);
            p.setPartitionSize(20);
            p.setKey();

            PreparedStatement insert = db.prepareStatement(insertQuery);

            insert.setLong(1, p.getId());
            insert.setLong(2, p.getTargetId());
            insert.setTimestamp(3, Timestamp.valueOf(p.getPartitionTs()));
            insert.executeUpdate();
            insert.close();

            out[i] = p;
        }
        return out;
    }

    @Test
    void testSaveGetDelete() throws SQLException {
        LocalDateTime batchTime = LocalDateTime.of(2020, 6, 1, 10, 0);
        var partitions = setupPartitions(batchTime);
        var depRes = new DependencyResult();
        depRes.setPartitionTs(batchTime);
        depRes.setPartitions(new Partition[]{partitions[0], partitions[1]});
        // TODO: Times set for these dependency partitions (and target id's) currently do not make much sense
        // but these are -not- checked when storing dependency partitions

        var dao = new PartitionDependencyDAO(db);
        var insertRes = dao.save(partitions[2].getId(), depRes, true);
        assertTrue(insertRes.isSuccess());
        assertFalse(insertRes.isError());

        insertRes = dao.save(partitions[2].getId(), depRes, false);
        assertFalse(insertRes.isSuccess());
        assertTrue(insertRes.isError());  // unable to overwrite will cause an insert error

        var allChildren = dao.getChildIds(partitions[0].getId());
        assertEquals(partitions[2].getId(), allChildren.get(0));

        var allParents = dao.getParentIds(partitions[2].getId());
        HashSet<Long> parentIds = new HashSet<>();
        parentIds.add(partitions[0].getId());
        parentIds.add(partitions[1].getId());
        for (Long l : allParents) {
            assertTrue(parentIds.contains(l));
        }

        var deleteRes = dao.remove(partitions[2].getId(), depRes);
        assertTrue(deleteRes.isSuccess());
        assertFalse(deleteRes.isError());

        allChildren = dao.getChildIds(partitions[0].getId());
        assertEquals(0, allChildren.size());
        allParents = dao.getParentIds(partitions[2].getId());
        assertEquals(0, allParents.size());
    }

    @Test
    void testMultiple() throws SQLException {
        LocalDateTime batchTime = LocalDateTime.of(2020, 6, 1, 10, 0);
        var partitions = setupPartitions(batchTime);
        var depResA = new DependencyResult();
        depResA.setPartitionTs(batchTime);
        depResA.setPartitions(new Partition[]{partitions[0]});
        var depResB = new DependencyResult();
        depResB.setPartitionTs(batchTime);
        depResB.setPartitions(new Partition[]{partitions[1]});

        var dao = new PartitionDependencyDAO(db);
        var insertRes = dao.saveAll(partitions[2].getId(),
                new DependencyResult[]{depResA, depResB}, false);
        assertTrue(insertRes.isSuccess());
        assertFalse(insertRes.isError());

        insertRes = dao.saveAll(partitions[2].getId(),
                new DependencyResult[]{depResA, depResB}, false);
        assertFalse(insertRes.isSuccess());
        assertTrue(insertRes.isError());

        var deleteRes = dao.removeAll(partitions[2].getId());

        insertRes = dao.saveAll(partitions[2].getId(),
                new DependencyResult[]{depResA, depResB}, false);
        assertTrue(insertRes.isSuccess());
        assertFalse(insertRes.isError());
    }

    @Test
    void gettingPartitions() throws SQLException {
        LocalDateTime batchTime = LocalDateTime.of(2020, 6, 1, 10, 0);
        var partitions = setupPartitions(batchTime);
        var depResA = new DependencyResult();
        depResA.setPartitionTs(batchTime);
        depResA.setPartitions(new Partition[]{partitions[0]});
        var dao = new PartitionDependencyDAO(db);
        var insertRes = dao.save(partitions[2].getId(), depResA, false);
        assertTrue(insertRes.isSuccess());
        assertFalse(insertRes.isError());

        var depResB = new DependencyResult();
        depResB.setPartitionTs(batchTime);
        depResB.setPartitions(new Partition[]{partitions[1]});
        var insertRes2 = dao.save(partitions[2].getId(), depResB, false);
        assertTrue(insertRes2.isSuccess());
        assertFalse(insertRes2.isError());

        var parentPartitions = dao.getParentPartitions(partitions[2]);
        assertEquals(2, parentPartitions.size());
    }

}

File: uhrwerk-backend-mysql/src/test/kotlin/io/qimia/uhrwerk/dao/PartitionDependencyDAOTest.kt
package io.qimia.uhrwerk.dao

import com.google.common.truth.Truth
import io.qimia.uhrwerk.TestData
import TestUtils
import io.qimia.uhrwerk.common.metastore.config.PartitionDependencyService
import io.qimia.uhrwerk.common.metastore.dependency.DependencyResult
import io.qimia.uhrwerk.common.metastore.model.ConnectionModel
import io.qimia.uhrwerk.common.metastore.model.Partition
import io.qimia.uhrwerk.common.metastore.model.TableModel
import io.qimia.uhrwerk.common.model.TargetModel
import io.qimia.uhrwerk.repo.*
import org.junit.jupiter.api.*
import org.slf4j.LoggerFactory
import org.testcontainers.containers.MySQLContainer
import org.testcontainers.containers.output.Slf4jLogConsumer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers
import java.time.LocalDateTime

@Testcontainers
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
internal class PartitionDependencyDAOTest {

    private val service: PartitionDependencyService = PartitionDependencyDAO()

    private var table: TableModel? = null
    private var connection: ConnectionModel? = null
    private var target: TargetModel? = null

    private var timestamps: List<LocalDateTime>? = null
    private var partitions: List<Partition>? = null


    @AfterEach
    fun cleanUp() {
        TestUtils.cleanData("PARTITION_DEPENDENCY", LOGGER)
        TestUtils.cleanData("PARTITION_", LOGGER)
        TestUtils.cleanData("TARGET", LOGGER)
        TestUtils.cleanData("TABLE_", LOGGER)
        TestUtils.cleanData("CONNECTION", LOGGER)
    }

    @BeforeEach
    fun saveData() {
        connection = ConnectionRepo().save(TestData.connection("Connection-PartitionTest"))
        table = TableRepo().save(TestData.table("Table-PartitionTest"))
        target = TargetRepo().save(TestData.target(connection!!.id!!, table!!.id!!))

        timestamps = TestData.timestamps(15)

        val parts = timestamps!!.map {
            TestData.partition(
                targetId = target!!.id!!,
                partitionTs = it
            )
        }
        val partRes = PartitionDAO().save(parts, true)
        partitions = partRes.map { it.newResult }
    }

    @Test
    @Order(1)
    fun save() {
        val batchTs = timestamps!!.last()
        val childPart = partitions!!.first()
        val parentParts = partitions!!.slice(0 until partitions!!.size - 1)

        val dependencyResult = DependencyResult()
        dependencyResult.partitionTs = batchTs
        dependencyResult.partitions = parentParts.toTypedArray()

        val result = service.saveAll(childPart.id, arrayOf(dependencyResult), false)

        Truth.assertThat(result.isSuccess).isTrue()
        Truth.assertThat(result.isError).isFalse()
    }

    @Test
    @Order(2)
    fun saveOverwrite() {

        val batchTs = timestamps!!.last()
        val childPart = partitions!!.first()
        val parentParts = partitions!!.slice(0 until partitions!!.size - 1)

        val dependencyResult = DependencyResult()
        dependencyResult.partitionTs = batchTs
        dependencyResult.partitions = parentParts.toTypedArray()

        service.saveAll(childPart.id, arrayOf(dependencyResult), true)

        val result = service.saveAll(childPart.id, arrayOf(dependencyResult), true)

        Truth.assertThat(result.isSuccess).isTrue()
        Truth.assertThat(result.isError).isFalse()
    }

    @Test
    @Order(3)
    fun saveFailNotOverwrite() {
        val batchTs = timestamps!!.last()
        val childPart = partitions!!.first()
        val parentParts = partitions!!.slice(0 until partitions!!.size - 1)

        val dependencyResult = DependencyResult()
        dependencyResult.partitionTs = batchTs
        dependencyResult.partitions = parentParts.toTypedArray()

        service.saveAll(childPart.id, arrayOf(dependencyResult), true)

        val result = service.saveAll(childPart.id, arrayOf(dependencyResult), false)

        Truth.assertThat(result.isSuccess).isFalse()
        Truth.assertThat(result.isError).isTrue()
    }


    companion object {
        private val LOGGER = LoggerFactory.getLogger(PartitionDependencyDAOTest::class.java)

        @Container
        var MY_SQL_DB: MySQLContainer<*> = TestUtils.mysqlContainer()

        @BeforeAll
        @JvmStatic
        fun setUp() {
            val logConsumer = Slf4jLogConsumer(LOGGER)
            MY_SQL_DB.followOutput(logConsumer)
            HikariCPDataSource.initConfig(
                MY_SQL_DB.jdbcUrl,
                MY_SQL_DB.username,
                MY_SQL_DB.password
            )
        }

        @AfterAll
        @JvmStatic
        fun tearDown() {
            HikariCPDataSource.close()
        }
    }
}